### ISSUE TITLE
Idea: AttributePropertyNeurons

### ISSUE THREAD

**Issue by ajthinking** [pipe-dream/core#6]
The core and a fileFactory selected as the "master" fileFactory together spawns a set of AttributePropertyNeurons:

```
let AttributePropertyNeurons = [
    // language agnostic things like datatype, unique, index
    defaultNeurons(),

    // language specific things like fillable for Laravel
    masterFileFactory.neurons()
]
```

```
Product
description > text:200, ..., ...
```

When a property has attached arguments like above it will test against <i>all</i> available neurons. If none, or more than one neuron fires it will throw a AmbiguityError. Else, if there is exactly one match that neuron will handle the provided argument and later down the line convert that to one or more keys in the schema.

By this approach the order of arguments wont matter and we could also implement a very forgiving syntax. Each Neuron could provide a class to do its thing. Example (in LaravelFileFactory):

```
FillableNeuron()
    fires(argument) {
        if argument == 'fillable' return true
        ...
    }

    schemaKey() {
        ...
        return {
            fillable: true (or false)
        }
    }
```

**Comment by septIO** [comment]
I'm not sure I'm following you here. Wouldn't a lexer be better than an AI? I'm all for reducing user inputs, and maximizing customization, but these has to collide at some point. I think having an AI would introduce too much `magic`. I think it would be easier for devs (who are the ones who are gonna use this tool, not end users) to have a general syntax to follow, and expect consistent results instead of relying on AIs to translate their key strokes. I might be wrong about this, or I might've completely misunderstood this issue.

**Comment by ajthinking** [comment]
Thats some good points for making a lexer. A lexer would provide a fixed syntax which is stable and predictable. Is this correct interpretion of the idea?
```
ATTRIBUTE [ > DATABASE_TYPE[optional:list:of:sub:args] [, RESERVED_KEYWORD:any:nbr:of:args]*
```
where `[] = optional` and `* = zero or more occurances`?

Please confirm or edit the row signature and I will get back with some more input! Thanks for the discussion!

**Comment by septIO** [comment]
> `ATTRIBUTE [ > DATABASE_TYPE[optional:list:of:sub:args] [, RESERVED_KEYWORD:any:nbr:of:args]*]`

That is correct.

**Comment by ajthinking** [comment]
So If I want to use a `RESERVED_KEYWORD` argument then I have to also supply `DATABASE_TYPE` as it cant be omitted?

Omitting sub arguments to use defaults requires the user to memorize the order (in this case the third position) `color > ::yellow`. Its secure, but might be hard to grasp for users. 

What about other things that are related to the database column like `index`, `unique`, `nullable`. Should they also be attached to the DATABASE_TYPE parameter group or should they be independent? If so it makes for a lot of :::::::. And how to make the distinction where they belonged?

Whats your thoughts about keyed arguments? `color > default:yellow` Its longer but slightly more intuitive.

In general I think we would benefit of finding a mix between strict/magic as they both comes with pros/cons.

**Comment by ajthinking** [comment]
Thought: by introducing a keyword for the database we could reduce the row signature to this:
```
ATTRIBUTE [ > RESERVED_KEYWORD:any:nbr:of:args]* ]
```
This would mean a few more keystrokes but the order does not matter.

Example
```
Car
speed > type:float:10:2, index:true
```

Keywords could also have defaults, like this:
```
Car
speed > index
```
The above would equal `index:true`

**Comment by septIO** [comment]
I think we mean the same thing here.
```
ATTR [ > [ANY_KEYWORDS[:ANY:ARG:FOR:KEYWORD]*]*
```

Meaning that
```
speed > float:10:2, index:true, unique:true, nullable:false, default:99.9
speed > float::2, index, unique, nullable:false, default:99.9
speed > float, index, unique, default:99.9
speed > float:::99.9, index, unique
```
All of the above examples deserializes to the exact same object.
**Assumptions**

Keyword | Assumption
--- | ---
index | false
unique | false
nullable | false
increment | false (only available on ints)

Attribute | Assumptions
--- | ---
boolean | length:1
bit | length: 2
tinyint | length: 3
smallint | length: 5
mediumint | length: 7
int | length:11
bigint | length: 19
float | length: 11, precision: 2
decimal | length: 11, precision: 2
char | length: 1
string | length: 250
text | length: 65535

table datatypes will be guessed by the github dump if the dev omits the datatype.
Basically, the verbosity can be small or big, depending on how strict the dev want it to be.
